### PR TITLE
Normalized temperature

### PR DESCRIPTION
**Pull request by kamillitman**

### CODE REVIEW

File: src/completions.rs [hunk 1]
```diff
@@ -15,27 +15,30 @@ pub struct Completions<T: LLMModel> {
...
-    ///
+    /// Constructor for the Completions API
     pub fn new(
         model: T,
         api_key: &str,
         max_tokens: Option<usize>,
         temperature: Option<u32>,
     ) -> Self {
+        let temperature = temperature
+            .map(|temp| model.get_normalized_temperature(temp))
```
Comment (kamillitman) [inline src/completions.rs:34]: This way we don't have to modify the constructor, making the change backwards compatible. We still pass an `u32` integer that should represent a 0-100[%] of the range we want to use. The `get_normalized_temperature` exposed by each model normalizes that to the range specific to the model.

File: src/llm_models/anthropic.rs [hunk 2]
```diff
@@ -80,7 +80,7 @@ impl LLMModel for AnthropicModels {
         json_schema: &Value,
         function_call: bool,
         max_tokens: &usize,
-        temperature: &u32,
+        temperature: &f32,
```
Comment (kamillitman) [inline src/llm_models/anthropic.rs:83]: This is never externally exposed so should be a safe change. We only set it with the constructor that hasn't changed.

File: src/llm_models/llm_model.rs [hunk 3]
```diff
@@ -69,4 +70,16 @@ pub trait LLMModel {
         //To be safe we go with smaller of the numbers
         std::cmp::min(max_requests_from_rpm, max_requests_from_tpm)
     }
+    ///Returns the default temperature to be used by the model
+    fn get_default_temperature(&self) -> f32 {
+        0f32
```
Comment (kamillitman) [inline src/llm_models/llm_model.rs:75]: default to 0 unless otherwise specified by the model

File: src/llm_models/llm_model.rs [hunk 4]
```diff
@@ -69,4 +70,16 @@ pub trait LLMModel {
...
+    }
+    ///Returns the normalized temperature for the model
+    //Input should be a 0-100 number representing the percentage of max temp for the model
+    fn get_normalized_temperature(&self, relative_temp: u32) -> f32 {
+        //Assuming 0-1 range for most models. Different ranges require model-specific implementations.
+        let min = 0u32;
```
Comment (kamillitman) [inline src/llm_models/llm_model.rs:81]: most models have a 0-1 range for temperature so this default implementation addresses most of them.

File: src/llm_models/openai.rs [hunk 5]
```diff
@@ -440,6 +440,14 @@ impl LLMModel for OpenAIModels {
             },
         }
     }
+
+    // Accepts a [0-100] percentage range and returns the target temperature based on model ranges
+    fn get_normalized_temperature(&self, relative_temp: u32) -> f32 {
+        // Temperature range documentation: https://platform.openai.com/docs/api-reference/chat/create
+        let min = 0u32;
+        let max = 2u32;
```
Comment (kamillitman) [inline src/llm_models/openai.rs:448]: OpenAI is the only one I found to have a 0-2 range, hence a model-specific implementaiton.

File: src/llm_models/llm_model.rs (hunk 3 above)
Comment (darioalessandro) [reply src/llm_models/llm_model.rs:75]: ```ChatGPT generates text through a probabilistic language modeling approach, where it writes responses word by word, calculating the most likely next word in the sequence. A key feature that influences this behavior is called temperature [[8](https://pmc.ncbi.nlm.nih.gov/articles/PMC10960206/#ref8), [9](https://pmc.ncbi.nlm.nih.gov/articles/PMC10960206/#ref9)]. In this context, temperature is a value from 0 to 2 that adjusts how random each subsequent word in the chat output is. A value of 0 will give the most probable word and, thus, the least variability. As the value increases toward and beyond 1, the next word becomes less probable, leading to more randomness and “creativity” in the response```

File: src/utils.rs [hunk 6]
```diff
@@ -55,4 +66,43 @@ mod tests {
...
+    #[test]
+    fn test_target_in_middle() {
+        assert_eq!(map_to_range(0, 100, 50), 50.0);
+        assert_eq!(map_to_range(10, 20, 50), 15.0);
+        assert_eq!(map_to_range(0, 1, 50), 0.5);
+    }
+
+    #[test]
+    fn test_target_out_of_bounds() {
+        assert_eq!(map_to_range(0, 100, 3000), 100.0); // Cap to 100
```
Comment (darioalessandro) [inline src/utils.rs:92]: I would throw an error if temp is outside of range, else users might feel that temp is being set correctly, what do you think?

File: Cargo.toml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "allms"
-version = "0.8.2"
+version = "0.8.3"
```
Comment (darioalessandro) [inline Cargo.toml:3]: pros: next time they do cargo `upgrade` they'll get this version. 

cons: the new range might be considered an api change, so technically we should bump minor at least?

X stands for a major version. The leftmost number denotes a major version. When you increase the major version number, you increase it by one but you reset both patch version and minor versions to zero. If the current version is 2.6.9 then the next upgrade for a major version will be 3.0.0. Increase the value of X when breaking the existing API.

Y stands for a minor version. It is used for the release of new functionality in the system. When you increase the minor version, you increase it by one but you must reset the patch version to zero. If the current version is 2.6.9 then the next upgrade for a minor version will be 2.7.0. Increase the value of Y when implementing new features in a backward-compatible way.

Z stands for a Patch Versions: Versions for patches are used for bug fixes. There are no functionality changes in the patch version upgrades. If the current version is 2.6.9 then the next version for a patch upgrade will be 2.6.10. There is no limit to these numbers. Increase the value of Z when fixing bugs

**Review by darioalessandro — APPROVED** [review]
lgtm but added some discussion regarding bumping minor vs patch version

File: src/utils.rs (hunk 6 above)
Comment (kamillitman) [reply src/utils.rs:92]: Check out the implementation of `utils:: map_to_range`" 
```
// Cap the target to the percentage range [0, 100]
let capped_target = target.min(100);

// Calculate the target value in the range [min, max]
let range = max as f32 - min as f32;
let percentage = capped_target as f32 / 100.0;
min as f32 + (range * percentage)
```
It actually caps the input to 100. So it ensures that we always set the temp correctly. It is guaranteed to be within the min-max range.

File: Cargo.toml (hunk 7 above)
Comment (kamillitman) [reply Cargo.toml:3]: Theoretically we are adding some new capabilities, but it's all done if a very backwards compatible way so there is absolutely no breaking changes here. People can safely upgrade without any worry.

Also, the new capabilities are not necessarily new features there are more enhancements of the existing feature. Previously you could set the temp directly in constructor, this just offers more control that the temp is in range for different providers.

Given this all, I would probably say it's more of patch of existing capabilities than something that would drive new uses.

File: src/llm_models/llm_model.rs (hunk 3 above)
Comment (kamillitman) [reply src/llm_models/llm_model.rs:75]: Correct. In the context of the crate, when we are using LLMs for generating type-safe responses, the default of more predictable and less creative seems reasonable, right?

File: Cargo.toml (hunk 7 above)
Comment (kamillitman) [reply Cargo.toml:3]: Alright, hold that thought. Turns out we were currently not setting the temperature for Assistants. I need to add this capability, which will probably mean a new minor. More changes coming to this PR....

Comment (kamillitman) [reply Cargo.toml]: Ok, I pushed some more changes that add the temperature setting for `Assistants` as well. This probably qualifies as new capabilities so I bumped the minor version to `0.9.0`.

File: src/assistants/openai/openai_assistant.rs [hunk 8]
```diff
@@ -48,6 +49,7 @@ impl OpenAIAssistant {
             id: None,
             thread_id: None,
             run_id: None,
+            temperature: model.get_default_temperature(),
```
Comment (kamillitman) [inline src/assistants/openai/openai_assistant.rs:52]: Turns out we weren't even passing temperature when creating an assistant and where therefore always using the default (which is set to 1). With the changes in this file we now actually set the temperature and give users the ability to define it as they see fit.

File: src/assistants/openai/openai_assistant.rs [hunk 9]
```diff
@@ -75,6 +77,25 @@ impl OpenAIAssistant {
         self
     }
 
+    ///
+    /// This method can be used to define the model temperature used by the Assistant
+    /// This method accepts % target of the acceptable range for the model
+    ///
+    pub fn temperature(mut self, temp_target: u32) -> Self {
```
Comment (kamillitman) [inline src/assistants/openai/openai_assistant.rs:84]: I gave users 2 options: either pass the temperature unchecked or use a safe way of setting it within the accepted range.

File: src/completions.rs [hunk 10]
```diff
@@ -15,54 +15,76 @@ pub struct Completions<T: LLMModel> {
...
-    /*
-     * This method can be used to provide values that will be used as context for the prompt.
-     * Using this function you can provide multiple input values by calling it multiple times. New values will be appended with the category name
-     * It accepts any instance that implements the Serialize trait.
-     */
+    ///
+    /// This method can be used to define the model temperature used by the Assistant
+    /// This method accepts % target of the acceptable range for the model
+    ///
+    pub fn temperature(mut self, temp_target: u32) -> Self {
```
Comment (kamillitman) [inline src/completions.rs:68]: Added the same methods for `Completions` as for `Assistants` to keep the interfaces consistent.

**Review by darioalessandro — APPROVED** [review]

File: src/llm_models/llm_model.rs (hunk 3 above)
Comment (darioalessandro) [reply src/llm_models/llm_model.rs:75]: yes

**Pull request merged**